harper-core/src/token_kind.rs

```diff
@@ -26,6 +27,49 @@ pub enum TokenKind {
     Regexish,
 }
 
+impl Hash for TokenKind {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        match self {
+            TokenKind::Word(metadata) => {
+                metadata.hash(state);
+            }
+            TokenKind::Punctuation(punct) => {
+                punct.hash(state);
+            }
+            TokenKind::Decade => {
+                0.hash(state);
+            }
+            TokenKind::Number(number) => {
+                number.hash(state);
+            }
+            TokenKind::Space(space) => {
+                space.hash(state);
+            }
+            TokenKind::Newline(newline) => {
+                newline.hash(state);
+            }
+            TokenKind::EmailAddress => {
+                0.hash(state);
+            }
+            TokenKind::Url => {
+                0.hash(state);
+            }
+            TokenKind::Hostname => {
+                0.hash(state);
+            }
+            TokenKind::Unlintable => {
+                0.hash(state);
+            }
+            TokenKind::ParagraphBreak => {
+                0.hash(state);
+            }
+            TokenKind::Regexish => {
+                0.hash(state);
+            }
```

Comment:
These hashes seem highly dubious to me. Still work in progress?

Reply:
> These hashes seem highly dubious to me. Still work in progress?

I believe @elijah-potter added that with a plan in mind but so far those are not really used by anything.

Also worth noting, the field the hashes are written into is not checked to see if something is already there and the new one stomps the old one. I wrote a patch that's in a PR to use a set instead.

Reply:
To be clear: my comment was in reference to the fact most variants have the same hash, which defeats the purpose of hashing.

> Also worth noting, the field the hashes are written into is not checked to see if something is already there and the new one stomps the old one. I wrote a patch that's in a PR to use a set instead.

Wdym by "the field"? This function is generic over the hasher, so we don't know anything about how hashing is done.

Reply:
Unless I'm on the wrong track and there's two kinds/places with hashing, the hash gets stored in a field in the metadata, `derived_from`.

Reply:
Ah, so you were talking about `WordMetadata::derived_from` and not about this hashing function. I thought you were talking about the `Hasher` implementation used in this function. 

And the hash in `WordMetadata::derived_from` is created by `WordId`, which completely unrelated this hashing function.